src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -34,24 +34,46 @@ public UdfComplexTypesTests(SparkFixture fixture)
         [Fact]
         public void TestUdfWithArrayType()
         {
-            // UDF with ArrayType throws a following exception:
-            // [] [] [Error] [TaskRunner] [0] ProcessStream() failed with exception: System.InvalidCastException: Unable to cast object of type 'System.Collections.ArrayList' to type 'System.Int32[]'.
-            //  at Microsoft.Spark.Sql.PicklingUdfWrapper`2.Execute(Int32 splitIndex, Object[] input, Int32[] argOffsets) in Microsoft.Spark\Sql\PicklingUdfWrapper.cs:line 44
-            //  at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.SingleCommandRunner.Run(Int32 splitId, Object input) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 239
-            //  at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 139
-            Func<Column, Column> udf = Udf<int[], string>(array => string.Join(',', array));
-            Assert.Throws<Exception>(() => _df.Select(udf(_df["ids"])).Show());
-
-            // Currently, there is a workaround to support ArrayType using ArrayList.
-            Func<Column, Column> workingUdf = Udf<ArrayList, string>(
-                array => string.Join(',', array.ToArray()));
-
-            Row[] rows = _df.Select(workingUdf(_df["ids"])).Collect().ToArray();
-            Assert.Equal(3, rows.Length);
+            // Simple Array.
+            {
+                Func<Column, Column> udf = Udf<int[], string>(
+                    array => string.Join(',', array.ToArray()));
+
+                Row[] rows = _df.Select(udf(_df["ids"])).Collect().ToArray();
+                Assert.Equal(3, rows.Length);
+
+                var expected = new string[] { "1", "3,5", "2,4" };
+                string[] rowsToArray = rows.Select(x => x[0].ToString()).ToArray();
+                Assert.Equal(expected, rowsToArray);
+            }
+
+            // Array of Arrays.
+            {
+                Func<Column, Column> udf = Udf<double[][], double>(

```

Comment:
can you pass in multiple arrays ?

Udf<double, double[], double[][], double[][][], double>  ?

Reply:
Added tests for multiple arrays. Working on `ArrayList` cases as mentioned.

Reply:
Do we also want to support the option of return value as `ArrayList`?
Like `Udf<double, ArrayList, ... , ArrayList>`?

@imback82 ideas?

Reply:
I guess it can be useful if you want to chain with Udf that takes in `ArrayList` (just for consistency).

Reply:
Got it, thanks!
 
The current logic is to cast `unpickledItems` to typed array if it contains `ArrayList` [here](https://github.com/dotnet/spark/blob/46b0d77d11fa916f92ad23d15cde794ad4ca5a68/src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs#L137). Another way might be casting input(after unpickling) [here](https://github.com/dotnet/spark/blob/46b0d77d11fa916f92ad23d15cde794ad4ca5a68/src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs#L152) when executing udf functions?

I am thinking either to pass [commands](https://github.com/dotnet/spark/blob/46b0d77d11fa916f92ad23d15cde794ad4ca5a68/src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs#L100) in `GetUnpickledObjects` or switch to the latter method, if we want to support `ArrayList` like `Udf<ArrayList, ArrayList, double>`. Please  suggest. Thanks.